shakenfist/baseobject.py

```diff
@@ -0,0 +1,105 @@
+
+from functools import partial
+import time
+
+from shakenfist import db
+from shakenfist import etcd
+from shakenfist import exceptions
+from shakenfist import logutil
+
+
+LOG, _ = logutil.setup(__name__)
+
+
+class DatabaseBackedObject(object):
+    object_type = 'unknown'
+
+    def __init__(self, object_uuid, version=None):
+        self.__uuid = object_uuid
+        self.__version = version
+
+    @property
+    def uuid(self):
+        return self.__uuid
+
+    @property
+    def version(self):
+        return self.__version
+
+    def __str__(self):
+        return '%s(%s)' % (type(self).object_type, self.__uuid)
+
+    def unique_label(self):
+        return (type(self).object_type, self.__uuid)
+
+    def add_event(self, operation, phase, duration=None, msg=None):
+        db.add_event(
+            type(self).object_type, self.__uuid, operation, phase, duration, msg)
+
+    @classmethod
+    def _db_create(cls, object_uuid, metadata):
+        metadata['uuid'] = object_uuid
+        etcd.create(cls.object_type, None, object_uuid, metadata)
+
+    @classmethod
+    def _db_get(cls, object_uuid):
+        # NOTE(mikal): we don't do upgrades inflight. They are assumed to have been
+        # done as part of the upgrade process.
+        o = etcd.get(cls.object_type, None, object_uuid)
+        if not o:
+            LOG.withField(cls.object_type, object_uuid).error(
+                'Object missing')
+            return None
+
+        if o.get('version', 0) != 2:
+            raise exceptions.BadObjectVersion(
+                'Unknown version - %s: %s' % (cls.object_type, o))
+        return o
+
+    def _db_get_attribute(self, attribute):
+        retval = etcd.get('attribute/%s' % type(self).object_type, self.__uuid, attribute)
+        if not retval:
+            return {}
+        return retval
+
+    def _db_set_attribute(self, attribute, value):
+        etcd.put('attribute/%s' % type(self).object_type, self.__uuid, attribute, value)
+
+    # Properties common to all objects which are routed to attributes
+    @property
+    def state(self):
+        return self._db_get_attribute('state')
+
+    def update_state(self, state, error_message=None):
```

Comment:
It would be good if `update_state` checked that the `state` is valid.

Reply:
As in the state exists in a whitelist, or that the movement from the original state to this state is valid?

Reply:
Both. :-)

Reply:
So I agree, but I think it should be done in a separate change, this one is getting too big.